rust/src/dns/log.rs

```diff
@@ -0,0 +1,212 @@
+/* Copyright (C) 2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+extern crate libc;
+
+use std;
+use std::string::String;
+
+use json::*;
+use dns::dns::*;
+use log::*;
+
+pub fn dns_rrtype_string(rrtype: u16) -> String {
+    match rrtype {
+        DNS_RTYPE_A => "A",
+        DNS_RTYPE_CNAME => "CNAME",
+        DNS_RTYPE_SOA => "SOA",
+        DNS_RTYPE_PTR => "PTR",
+        DNS_RTYPE_MX => "MX",
+        DNS_RTYPE_SSHFP => "SSHFP",
+        DNS_RTYPE_RRSIG => "RRSIG",
+        _ => {
+            return rrtype.to_string();
+        }
+    }.to_string()
+}
+
+fn dns_rcode_string(flags: u16) -> String {
+    match flags & 0x000f {
+        DNS_RCODE_NOERROR => "NOERROR",
+        DNS_RCODE_FORMERR => "FORMERR",
+        DNS_RCODE_NXDOMAIN => "NXDOMAIN",
+        _ => {
+            return (flags & 0x000f).to_string();
+        }
+    }.to_string()
+}
+
+/// Format bytes as an IP address string.
+pub fn dns_print_addr(addr: &Vec<u8>) -> std::string::String {
+    if addr.len() == 4 {
+        return format!("{}.{}.{}.{}", addr[0], addr[1], addr[2], addr[3]);
+    }
+    else if addr.len() == 16 {
+        return format!("{}{}:{}{}:{}{}:{}{}:{}{}:{}{}:{}{}:{}{}",
+                       addr[0],
+                       addr[1],
+                       addr[2],
+                       addr[3],
+                       addr[4],
+                       addr[5],
+                       addr[6],
+                       addr[7],
+                       addr[8],
+                       addr[9],
+                       addr[10],
+                       addr[11],
+                       addr[12],
+                       addr[13],
+                       addr[14],
+                       addr[15]);
+    }
+    else {
+        return "".to_string();
+    }
+}
+
+///  Log the SSHPF in an DNSAnswerEntry.
+fn dns_log_sshfp(js: &Json, answer: &DNSAnswerEntry)
+{
+    // Need at least 3 bytes - TODO: log something if we don't?
```

Comment:
I'm not sure if events should be logged from the logger - if logging is turned off it will never be logged. That would mean moving validation into post-parsing. Since it parse fully valid, but still not be a valid sshfp.

Here's a thought.  If we can't log it, maybe do something like:
```
{
    ...
    "sshfp": {
        "invalid": true
    }
}
```

Or even just log it as empty...?